File: lib/features/race_session/race_session_models.dart
import 'dart:convert';

enum SessionStage { setup, lobby, monitoring }

enum SessionNetworkRole { none, host, client }

enum SessionDeviceRole { unassigned, start, split, stop }

enum SessionCameraFacing { rear, front }

class SessionDevice {
  const SessionDevice({
    required this.id,
    required this.name,
    required this.role,
    this.cameraFacing = SessionCameraFacing.rear,
    this.highSpeedEnabled = false,
    required this.isLocal,
  });

  final String id;
  final String name;
  final SessionDeviceRole role;
  final SessionCameraFacing cameraFacing;
  final bool highSpeedEnabled;
  final bool isLocal;

  SessionDevice copyWith({
    String? id,
    String? name,
    SessionDeviceRole? role,
    SessionCameraFacing? cameraFacing,
    bool? highSpeedEnabled,
    bool? isLocal,
  }) {
    return SessionDevice(
      id: id ?? this.id,
      name: name ?? this.name,
      role: role ?? this.role,
      cameraFacing: cameraFacing ?? this.cameraFacing,
      highSpeedEnabled: highSpeedEnabled ?? this.highSpeedEnabled,
      isLocal: isLocal ?? this.isLocal,
    );
  }

  Map<String, dynamic> toJson() {
    return <String, dynamic>{
      'id': id,
      'name': name,
      'role': role.name,
      'cameraFacing': cameraFacing.name,
      'highSpeedEnabled': highSpeedEnabled,
      'isLocal': isLocal,
    };
  }

  static SessionDevice? fromJson(dynamic source) {
    if (source is! Map<String, dynamic>) {
      return null;
    }
    final id = source['id']?.toString();
    final name = source['name']?.toString();
    final role = sessionDeviceRoleFromName(source['role']?.toString());
    final cameraFacing =
        sessionCameraFacingFromName(source['cameraFacing']?.toString()) ??
        SessionCameraFacing.rear;
    final highSpeedEnabled = source['highSpeedEnabled'] == true;
    if (id == null || id.isEmpty || name == null || role == null) {
      return null;
    }
    return SessionDevice(
      id: id,
      name: name,
      role: role,
      cameraFacing: cameraFacing,
      highSpeedEnabled: highSpeedEnabled,
      isLocal: source['isLocal'] == true,
    );
  }
}

class SessionRaceTimeline {
  const SessionRaceTimeline({
    this.startedSensorNanos,
    required this.splitElapsedNanos,
    this.stopElapsedNanos,
  });

  final int? startedSensorNanos;
  final List<int> splitElapsedNanos;
  final int? stopElapsedNanos;

  factory SessionRaceTimeline.idle() {
    return const SessionRaceTimeline(splitElapsedNanos: <int>[]);
  }

  bool get hasStarted => startedSensorNanos != null;
  bool get hasStopped => stopElapsedNanos != null;
  bool get isRunning => hasStarted && !hasStopped;

  SessionRaceTimeline copyWith({
    int? startedSensorNanos,
    List<int>? splitElapsedNanos,
    int? stopElapsedNanos,
    bool clearStartedSensor = false,
    bool clearStopElapsedNanos = false,
  }) {
    return SessionRaceTimeline(
      startedSensorNanos: clearStartedSensor
          ? null
          : (startedSensorNanos ?? this.startedSensorNanos),
      splitElapsedNanos: splitElapsedNanos ?? this.splitElapsedNanos,
      stopElapsedNanos: clearStopElapsedNanos
          ? null
          : (stopElapsedNanos ?? this.stopElapsedNanos),
    );
  }

  Map<String, dynamic> toJson() {
    return <String, dynamic>{
      'startedSensorNanos': startedSensorNanos,
      'splitElapsedNanos': splitElapsedNanos,
      'stopElapsedNanos': stopElapsedNanos,
    };
  }

  static SessionRaceTimeline fromJson(dynamic source) {
    if (source is! Map<String, dynamic>) {
      return SessionRaceTimeline.idle();
    }
    final splitRaw = source['splitElapsedNanos'];
    final splitElapsedNanos = <int>[];
    if (splitRaw is List) {
      for (final value in splitRaw) {
        if (value is int) {
          splitElapsedNanos.add(value);
        } else if (value is num) {
          splitElapsedNanos.add(value.toInt());
        }
      }
    }
    final startedRaw = source['startedSensorNanos'];
    final stopElapsedRaw = source['stopElapsedNanos'];
    final startedSensorNanos = startedRaw is num ? startedRaw.toInt() : null;
    final stopElapsedNanos = stopElapsedRaw is num
        ? stopElapsedRaw.toInt()
        : null;
    return SessionRaceTimeline(
      startedSensorNanos: startedSensorNanos,
      splitElapsedNanos: splitElapsedNanos,
      stopElapsedNanos: stopElapsedNanos,
    );
  }
}

class SessionSnapshotMessage {
  const SessionSnapshotMessage({
    required this.stage,
    required this.monitoringActive,
    required this.devices,
    required this.timeline,
    this.hostSensorMinusElapsedNanos,
    this.hostGpsUtcOffsetNanos,
    this.hostGpsFixAgeNanos,
    this.selfDeviceId,
  });

  final SessionStage stage;
  final bool monitoringActive;
  final List<SessionDevice> devices;
  final SessionRaceTimeline timeline;
  final int? hostSensorMinusElapsedNanos;
  final int? hostGpsUtcOffsetNanos;
  final int? hostGpsFixAgeNanos;
  final String? selfDeviceId;

  Map<String, dynamic> toJson() {
    return <String, dynamic>{
      'type': 'snapshot',
      'stage': stage.name,
      'monitoringActive': monitoringActive,
      'devices': devices.map((device) => device.toJson()).toList(),
      'timeline': timeline.toJson(),
      'hostSensorMinusElapsedNanos': hostSensorMinusElapsedNanos,
      'hostGpsUtcOffsetNanos': hostGpsUtcOffsetNanos,
      'hostGpsFixAgeNanos': hostGpsFixAgeNanos,
      'selfDeviceId': selfDeviceId,
    };
  }

  String toJsonString() => jsonEncode(toJson());

  static SessionSnapshotMessage? tryParse(String raw) {
    try {
      final decoded = jsonDecode(raw);
      if (decoded is! Map<String, dynamic> || decoded['type'] != 'snapshot') {
        return null;
      }
      final stage = sessionStageFromName(decoded['stage']?.toString());
      if (stage == null) {
        return null;
      }
      final monitoringActive = decoded['monitoringActive'] == true;
      final devicesRaw = decoded['devices'];
      if (devicesRaw is! List) {
        return null;
      }
      final devices = <SessionDevice>[];
      for (final item in devicesRaw) {
        final parsed = SessionDevice.fromJson(item);
        if (parsed != null) {
          devices.add(parsed);
        }
      }
      if (devices.isEmpty) {
        return null;
      }
      return SessionSnapshotMessage(
        stage: stage,
        monitoringActive: monitoringActive,
        devices: devices,
        timeline: SessionRaceTimeline.fromJson(decoded['timeline']),
        hostSensorMinusElapsedNanos:
            (decoded['hostSensorMinusElapsedNanos'] as num?)?.toInt(),
        hostGpsUtcOffsetNanos: (decoded['hostGpsUtcOffsetNanos'] as num?)
            ?.toInt(),
        hostGpsFixAgeNanos: (decoded['hostGpsFixAgeNanos'] as num?)?.toInt(),
        selfDeviceId: decoded['selfDeviceId']?.toString(),
      );
    } catch (_) {
      return null;
    }
  }
}

class SessionTriggerRequestMessage {
  const SessionTriggerRequestMessage({
    required this.role,
    required this.triggerSensorNanos,
    this.mappedHostSensorNanos,
  });

  final SessionDeviceRole role;
  final int triggerSensorNanos;
  final int? mappedHostSensorNanos;

  Map<String, dynamic> toJson() {
    return <String, dynamic>{
      'type': 'trigger_request',
      'role': role.name,
      'triggerSensorNanos': triggerSensorNanos,
      'mappedHostSensorNanos': mappedHostSensorNanos,
    };
  }

  String toJsonString() => jsonEncode(toJson());

  static SessionTriggerRequestMessage? tryParse(String raw) {
    try {
      final decoded = jsonDecode(raw);
      if (decoded is! Map<String, dynamic> ||
          decoded['type'] != 'trigger_request') {
        return null;
      }
      final role = sessionDeviceRoleFromName(decoded['role']?.toString());
      final triggerSensorNanosRaw = decoded['triggerSensorNanos'];
      if (role == null || triggerSensorNanosRaw is! num) {
        return null;
      }
      final mappedHostSensorNanosRaw = decoded['mappedHostSensorNanos'];
      return SessionTriggerRequestMessage(
        role: role,
        triggerSensorNanos: triggerSensorNanosRaw.toInt(),
        mappedHostSensorNanos: mappedHostSensorNanosRaw is num
            ? mappedHostSensorNanosRaw.toInt()
            : null,
      );
    } catch (_) {
      return null;
    }
  }
}

class SessionClockSyncRequestMessage {
  const SessionClockSyncRequestMessage({required this.clientSendElapsedNanos});

  final int clientSendElapsedNanos;

  Map<String, dynamic> toJson() {
    return <String, dynamic>{
      'type': 'clock_sync_request',
      'clientSendElapsedNanos': clientSendElapsedNanos,
    };
  }

  String toJsonString() => jsonEncode(toJson());

  static SessionClockSyncRequestMessage? tryParse(String raw) {
    try {
      final decoded = jsonDecode(raw);
      if (decoded is! Map<String, dynamic> ||
          decoded['type'] != 'clock_sync_request') {
        return null;
      }
      final clientSendElapsedNanosRaw = decoded['clientSendElapsedNanos'];
      if (clientSendElapsedNanosRaw is! num) {
        return null;
      }
      return SessionClockSyncRequestMessage(
        clientSendElapsedNanos: clientSendElapsedNanosRaw.toInt(),
      );
    } catch (_) {
      return null;
    }
  }
}

class SessionClockSyncResponseMessage {
  const SessionClockSyncResponseMessage({
    required this.clientSendElapsedNanos,
    required this.hostReceiveElapsedNanos,
    required this.hostSendElapsedNanos,
  });

  final int clientSendElapsedNanos;
  final int hostReceiveElapsedNanos;
  final int hostSendElapsedNanos;

  Map<String, dynamic> toJson() {
    return <String, dynamic>{
      'type': 'clock_sync_response',
      'clientSendElapsedNanos': clientSendElapsedNanos,
      'hostReceiveElapsedNanos': hostReceiveElapsedNanos,
      'hostSendElapsedNanos': hostSendElapsedNanos,
    };
  }

  String toJsonString() => jsonEncode(toJson());

  static SessionClockSyncResponseMessage? tryParse(String raw) {
    try {
      final decoded = jsonDecode(raw);
      if (decoded is! Map<String, dynamic> ||
          decoded['type'] != 'clock_sync_response') {
        return null;
      }
      final clientSendElapsedNanosRaw = decoded['clientSendElapsedNanos'];
      final hostReceiveElapsedNanosRaw = decoded['hostReceiveElapsedNanos'];
      final hostSendElapsedNanosRaw = decoded['hostSendElapsedNanos'];
      if (clientSendElapsedNanosRaw is! num ||
          hostReceiveElapsedNanosRaw is! num ||
          hostSendElapsedNanosRaw is! num) {
        return null;
      }
      return SessionClockSyncResponseMessage(
        clientSendElapsedNanos: clientSendElapsedNanosRaw.toInt(),
        hostReceiveElapsedNanos: hostReceiveElapsedNanosRaw.toInt(),
        hostSendElapsedNanos: hostSendElapsedNanosRaw.toInt(),
      );
    } catch (_) {
      return null;
    }
  }
}

SessionStage? sessionStageFromName(String? name) {
  if (name == null) {
    return null;
  }
  for (final value in SessionStage.values) {
    if (value.name == name) {
      return value;
    }
  }
  return null;
}

SessionDeviceRole? sessionDeviceRoleFromName(String? name) {
  if (name == null) {
    return null;
  }
  for (final value in SessionDeviceRole.values) {
    if (value.name == name) {
      return value;
    }
  }
  return null;
}

SessionCameraFacing? sessionCameraFacingFromName(String? name) {
  if (name == null) {
    return null;
  }
  for (final value in SessionCameraFacing.values) {
    if (value.name == name) {
      return value;
    }
  }
  return null;
}

String sessionDeviceRoleLabel(SessionDeviceRole role) {
  switch (role) {
    case SessionDeviceRole.unassigned:
      return 'Unassigned';
    case SessionDeviceRole.start:
      return 'Start';
    case SessionDeviceRole.split:
      return 'Split';
    case SessionDeviceRole.stop:
      return 'Stop';
  }
}

String sessionCameraFacingLabel(SessionCameraFacing facing) {
  switch (facing) {
    case SessionCameraFacing.rear:
      return 'Rear';
    case SessionCameraFacing.front:
      return 'Front';
  }
}

File: android/app/src/main/kotlin/com/paul/sprintsync/features/race_session/RaceSessionModels.kt
package com.paul.sprintsync.features.race_session

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

enum class SessionStage {
    SETUP,
    LOBBY,
    MONITORING,
}

enum class SessionNetworkRole {
    NONE,
    HOST,
    CLIENT,
}

enum class SessionDeviceRole {
    UNASSIGNED,
    START,
    SPLIT,
    STOP,
}

enum class SessionCameraFacing {
    REAR,
    FRONT,
}

data class SessionDevice(
    val id: String,
    val name: String,
    val role: SessionDeviceRole,
    val cameraFacing: SessionCameraFacing = SessionCameraFacing.REAR,
    val highSpeedEnabled: Boolean = false,
    val isLocal: Boolean,
) {
    fun toJsonObject(): JSONObject {
        return JSONObject()
            .put("id", id)
            .put("name", name)
            .put("role", role.name.lowercase())
            .put("cameraFacing", cameraFacing.name.lowercase())
            .put("highSpeedEnabled", highSpeedEnabled)
            .put("isLocal", isLocal)
    }

    companion object {
        fun fromJsonObject(decoded: JSONObject): SessionDevice? {
            val id = decoded.optString("id", "").trim()
            val name = decoded.optString("name", "").trim()
            val role = sessionDeviceRoleFromName(decoded.readOptionalString("role"))
            val cameraFacing = sessionCameraFacingFromName(decoded.readOptionalString("cameraFacing"))
                ?: SessionCameraFacing.REAR
            if (id.isEmpty() || name.isEmpty() || role == null) {
                return null
            }
            return SessionDevice(
                id = id,
                name = name,
                role = role,
                cameraFacing = cameraFacing,
                highSpeedEnabled = decoded.optBoolean("highSpeedEnabled", false),
                isLocal = decoded.optBoolean("isLocal", false),
            )
        }
    }
}

data class SessionSnapshotMessage(
    val stage: SessionStage,
    val monitoringActive: Boolean,
    val devices: List<SessionDevice>,
    val hostStartSensorNanos: Long?,
    val hostSplitSensorNanos: List<Long>,
    val hostStopSensorNanos: Long?,
    val runId: String?,
    val hostSensorMinusElapsedNanos: Long?,
    val hostGpsUtcOffsetNanos: Long?,
    val hostGpsFixAgeNanos: Long?,
    val selfDeviceId: String?,
) {
    fun toJsonString(): String {
        val devicesArray = JSONArray()
        devices.forEach { devicesArray.put(it.toJsonObject()) }
        val splitArray = JSONArray()
        hostSplitSensorNanos.forEach { splitArray.put(it) }
        val timeline = JSONObject()
            .put("hostStartSensorNanos", hostStartSensorNanos ?: JSONObject.NULL)
            .put("hostSplitSensorNanos", splitArray)
            .put("hostStopSensorNanos", hostStopSensorNanos ?: JSONObject.NULL)
        return JSONObject()
            .put("type", TYPE)
            .put("stage", stage.name.lowercase())
            .put("monitoringActive", monitoringActive)
            .put("devices", devicesArray)
            .put("timeline", timeline)
            .put("runId", runId ?: JSONObject.NULL)
            .put("hostSensorMinusElapsedNanos", hostSensorMinusElapsedNanos ?: JSONObject.NULL)
            .put("hostGpsUtcOffsetNanos", hostGpsUtcOffsetNanos ?: JSONObject.NULL)
            .put("hostGpsFixAgeNanos", hostGpsFixAgeNanos ?: JSONObject.NULL)
            .put("selfDeviceId", selfDeviceId ?: JSONObject.NULL)
            .toString()
    }

    companion object {
        const val TYPE = "snapshot"

        fun tryParse(raw: String): SessionSnapshotMessage? {
            val decoded = try {
                JSONObject(raw)
            } catch (_: JSONException) {
                return null
            }
            if (decoded.optString("type") != TYPE) {
                return null
            }
            val stage = sessionStageFromName(decoded.readOptionalString("stage")) ?: return null
            val devicesRaw = decoded.optJSONArray("devices") ?: return null
            val parsedDevices = mutableListOf<SessionDevice>()
            for (index in 0 until devicesRaw.length()) {
                val item = devicesRaw.optJSONObject(index) ?: continue
                val parsed = SessionDevice.fromJsonObject(item) ?: continue
                parsedDevices += parsed
            }
            if (parsedDevices.isEmpty()) {
                return null
            }
            val timeline = decoded.optJSONObject("timeline") ?: JSONObject()
            val splitRaw = timeline.optJSONArray("hostSplitSensorNanos") ?: JSONArray()
            val splits = mutableListOf<Long>()
            for (index in 0 until splitRaw.length()) {
                val value = splitRaw.optLong(index, Long.MIN_VALUE)
                if (value != Long.MIN_VALUE) {
                    splits += value
                }
            }
            return SessionSnapshotMessage(
                stage = stage,
                monitoringActive = decoded.optBoolean("monitoringActive", false),
                devices = parsedDevices,
                hostStartSensorNanos = timeline.readOptionalLong("hostStartSensorNanos"),
                hostSplitSensorNanos = splits,
                hostStopSensorNanos = timeline.readOptionalLong("hostStopSensorNanos"),
                runId = decoded.optString("runId", "").ifBlank { null },
                hostSensorMinusElapsedNanos = decoded.readOptionalLong("hostSensorMinusElapsedNanos"),
                hostGpsUtcOffsetNanos = decoded.readOptionalLong("hostGpsUtcOffsetNanos"),
                hostGpsFixAgeNanos = decoded.readOptionalLong("hostGpsFixAgeNanos"),
                selfDeviceId = decoded.optString("selfDeviceId", "").ifBlank { null },
            )
        }
    }
}

data class SessionTriggerRequestMessage(
    val role: SessionDeviceRole,
    val triggerSensorNanos: Long,
    val mappedHostSensorNanos: Long?,
) {
    fun toJsonString(): String {
        return JSONObject()
            .put("type", TYPE)
            .put("role", role.name.lowercase())
            .put("triggerSensorNanos", triggerSensorNanos)
            .put("mappedHostSensorNanos", mappedHostSensorNanos ?: JSONObject.NULL)
            .toString()
    }

    companion object {
        const val TYPE = "trigger_request"

        fun tryParse(raw: String): SessionTriggerRequestMessage? {
            val decoded = try {
                JSONObject(raw)
            } catch (_: JSONException) {
                return null
            }
            if (decoded.optString("type") != TYPE) {
                return null
            }
            val role = sessionDeviceRoleFromName(decoded.readOptionalString("role")) ?: return null
            val triggerSensorNanos = decoded.optLong("triggerSensorNanos", Long.MIN_VALUE)
            if (triggerSensorNanos == Long.MIN_VALUE) {
                return null
            }
            return SessionTriggerRequestMessage(
                role = role,
                triggerSensorNanos = triggerSensorNanos,
                mappedHostSensorNanos = decoded.readOptionalLong("mappedHostSensorNanos"),
            )
        }
    }
}

data class SessionTriggerRefinementMessage(
    val runId: String,
    val role: SessionDeviceRole,
    val provisionalHostSensorNanos: Long,
    val refinedHostSensorNanos: Long,
    val splitIndex: Int,
) {
    fun toJsonString(): String {
        return JSONObject()
            .put("type", TYPE)
            .put("runId", runId)
            .put("role", role.name.lowercase())
            .put("provisionalHostSensorNanos", provisionalHostSensorNanos)
            .put("refinedHostSensorNanos", refinedHostSensorNanos)
            .put("splitIndex", splitIndex)
            .toString()
    }

    companion object {
        const val TYPE = "trigger_refinement"

        fun tryParse(raw: String): SessionTriggerRefinementMessage? {
            val decoded = try {
                JSONObject(raw)
            } catch (_: JSONException) {
                return null
            }
            if (decoded.optString("type") != TYPE) {
                return null
            }
            val runId = decoded.optString("runId", "").trim()
            val role = sessionDeviceRoleFromName(decoded.readOptionalString("role"))
            val provisional = decoded.optLong("provisionalHostSensorNanos", Long.MIN_VALUE)
            val refined = decoded.optLong("refinedHostSensorNanos", Long.MIN_VALUE)
            val splitIndex = decoded.optInt("splitIndex", -1)
            if (runId.isEmpty() || role == null || provisional == Long.MIN_VALUE || refined == Long.MIN_VALUE || splitIndex < 0) {
                return null
            }
            return SessionTriggerRefinementMessage(
                runId = runId,
                role = role,
                provisionalHostSensorNanos = provisional,
                refinedHostSensorNanos = refined,
                splitIndex = splitIndex,
            )
        }
    }
}

data class SessionClockSyncRequestMessage(
    val clientSendElapsedNanos: Long,
) {
    fun toJsonString(): String {
        return JSONObject()
            .put("type", TYPE)
            .put("clientSendElapsedNanos", clientSendElapsedNanos)
            .toString()
    }

    companion object {
        const val TYPE = "clock_sync_request"

        fun tryParse(raw: String): SessionClockSyncRequestMessage? {
            val decoded = try {
                JSONObject(raw)
            } catch (_: JSONException) {
                return null
            }
            if (decoded.optString("type") != TYPE) {
                return null
            }
            if (!decoded.has("clientSendElapsedNanos")) {
                return null
            }
            val clientSendElapsedNanos = decoded.optLong("clientSendElapsedNanos", Long.MIN_VALUE)
            if (clientSendElapsedNanos == Long.MIN_VALUE) {
                return null
            }
            return SessionClockSyncRequestMessage(clientSendElapsedNanos = clientSendElapsedNanos)
        }
    }
}

data class SessionClockSyncResponseMessage(
    val clientSendElapsedNanos: Long,
    val hostReceiveElapsedNanos: Long,
    val hostSendElapsedNanos: Long,
) {
    fun toJsonString(): String {
        return JSONObject()
            .put("type", TYPE)
            .put("clientSendElapsedNanos", clientSendElapsedNanos)
            .put("hostReceiveElapsedNanos", hostReceiveElapsedNanos)
            .put("hostSendElapsedNanos", hostSendElapsedNanos)
            .toString()
    }

    companion object {
        const val TYPE = "clock_sync_response"

        fun tryParse(raw: String): SessionClockSyncResponseMessage? {
            val decoded = try {
                JSONObject(raw)
            } catch (_: JSONException) {
                return null
            }
            if (decoded.optString("type") != TYPE) {
                return null
            }
            val clientSend = decoded.optLong("clientSendElapsedNanos", Long.MIN_VALUE)
            val hostReceive = decoded.optLong("hostReceiveElapsedNanos", Long.MIN_VALUE)
            val hostSend = decoded.optLong("hostSendElapsedNanos", Long.MIN_VALUE)
            if (clientSend == Long.MIN_VALUE || hostReceive == Long.MIN_VALUE || hostSend == Long.MIN_VALUE) {
                return null
            }
            return SessionClockSyncResponseMessage(
                clientSendElapsedNanos = clientSend,
                hostReceiveElapsedNanos = hostReceive,
                hostSendElapsedNanos = hostSend,
            )
        }
    }
}

data class SessionTimelineSnapshotMessage(
    val hostStartSensorNanos: Long?,
    val hostSplitSensorNanos: List<Long>,
    val hostStopSensorNanos: Long?,
    val sentElapsedNanos: Long,
) {
    fun toJsonString(): String {
        val splits = JSONArray()
        hostSplitSensorNanos.forEach { splits.put(it) }
        return JSONObject()
            .put("type", TYPE)
            .put("hostStartSensorNanos", hostStartSensorNanos ?: JSONObject.NULL)
            .put("hostSplitSensorNanos", splits)
            .put("hostStopSensorNanos", hostStopSensorNanos ?: JSONObject.NULL)
            .put("sentElapsedNanos", sentElapsedNanos)
            .toString()
    }

    companion object {
        const val TYPE = "timeline_snapshot"

        fun tryParse(raw: String): SessionTimelineSnapshotMessage? {
            val decoded = try {
                JSONObject(raw)
            } catch (_: JSONException) {
                return null
            }
            if (decoded.optString("type") != TYPE) {
                return null
            }
            val sentElapsedNanos = decoded.optLong("sentElapsedNanos", Long.MIN_VALUE)
            if (sentElapsedNanos == Long.MIN_VALUE) {
                return null
            }
            val hostStartSensorNanos = decoded.readOptionalLong("hostStartSensorNanos")
            val hostStopSensorNanos = decoded.readOptionalLong("hostStopSensorNanos")
            val splits = decoded.optJSONArray("hostSplitSensorNanos") ?: JSONArray()
            val parsedSplits = mutableListOf<Long>()
            for (index in 0 until splits.length()) {
                val value = splits.optLong(index, Long.MIN_VALUE)
                if (value != Long.MIN_VALUE) {
                    parsedSplits += value
                }
            }
            return SessionTimelineSnapshotMessage(
                hostStartSensorNanos = hostStartSensorNanos,
                hostSplitSensorNanos = parsedSplits,
                hostStopSensorNanos = hostStopSensorNanos,
                sentElapsedNanos = sentElapsedNanos,
            )
        }
    }
}

data class SessionTriggerMessage(
    val triggerType: String,
    val splitIndex: Int,
    val triggerSensorNanos: Long,
) {
    fun toJsonString(): String {
        return JSONObject()
            .put("type", TYPE)
            .put("triggerType", triggerType)
            .put("splitIndex", splitIndex)
            .put("triggerSensorNanos", triggerSensorNanos)
            .toString()
    }

    companion object {
        const val TYPE = "session_trigger"

        fun tryParse(raw: String): SessionTriggerMessage? {
            val decoded = try {
                JSONObject(raw)
            } catch (_: JSONException) {
                return null
            }
            if (decoded.optString("type") != TYPE) {
                return null
            }
            val triggerType = decoded.optString("triggerType", "").trim()
            val splitIndex = decoded.optInt("splitIndex", -1)
            val triggerSensorNanos = decoded.optLong("triggerSensorNanos", Long.MIN_VALUE)
            if (triggerType.isEmpty() || splitIndex < 0 || triggerSensorNanos == Long.MIN_VALUE) {
                return null
            }
            return SessionTriggerMessage(
                triggerType = triggerType,
                splitIndex = splitIndex,
                triggerSensorNanos = triggerSensorNanos,
            )
        }
    }
}

data class SessionSwitchToP2pMessage(val timestampNanos: Long) {
    fun toJsonString(): String {
        return JSONObject()
            .put("type", TYPE)
            .put("timestampNanos", timestampNanos)
            .toString()
    }

    companion object {
        const val TYPE = "switch_to_p2p"

        fun tryParse(raw: String): SessionSwitchToP2pMessage? {
            val decoded = try {
                JSONObject(raw)
            } catch (_: JSONException) {
                return null
            }
            if (decoded.optString("type") != TYPE) {
                return null
            }
            return SessionSwitchToP2pMessage(
                timestampNanos = decoded.optLong("timestampNanos", 0L)
            )
        }
    }
}

data class SessionDeviceIdentityMessage(
    val stableDeviceId: String,
    val deviceName: String,
) {
    fun toJsonString(): String {
        return JSONObject()
            .put("type", TYPE)
            .put("stableDeviceId", stableDeviceId)
            .put("deviceName", deviceName)
            .toString()
    }

    companion object {
        const val TYPE = "device_identity"

        fun tryParse(raw: String): SessionDeviceIdentityMessage? {
            val decoded = try {
                JSONObject(raw)
            } catch (_: JSONException) {
                return null
            }
            if (decoded.optString("type") != TYPE) {
                return null
            }
            val stableDeviceId = decoded.optString("stableDeviceId", "").trim()
            val deviceName = decoded.optString("deviceName", "").trim()
            if (stableDeviceId.isEmpty() || deviceName.isEmpty()) {
                return null
            }
            return SessionDeviceIdentityMessage(
                stableDeviceId = stableDeviceId,
                deviceName = deviceName,
            )
        }
    }
}

fun sessionStageFromName(name: String?): SessionStage? {
    if (name == null) {
        return null
    }
    return SessionStage.values().firstOrNull { it.name.equals(name.trim(), ignoreCase = true) }
}

fun sessionDeviceRoleFromName(name: String?): SessionDeviceRole? {
    if (name == null) {
        return null
    }
    return SessionDeviceRole.values().firstOrNull { it.name.equals(name.trim(), ignoreCase = true) }
}

fun sessionCameraFacingFromName(name: String?): SessionCameraFacing? {
    if (name == null) {
        return null
    }
    return SessionCameraFacing.values().firstOrNull { it.name.equals(name.trim(), ignoreCase = true) }
}

fun sessionDeviceRoleLabel(role: SessionDeviceRole): String {
    return when (role) {
        SessionDeviceRole.UNASSIGNED -> "Unassigned"
        SessionDeviceRole.START -> "Start"
        SessionDeviceRole.SPLIT -> "Split"
        SessionDeviceRole.STOP -> "Stop"
    }
}

fun sessionCameraFacingLabel(facing: SessionCameraFacing): String {
    return when (facing) {
        SessionCameraFacing.REAR -> "Rear"
        SessionCameraFacing.FRONT -> "Front"
    }
}

private fun JSONObject.readOptionalLong(key: String): Long? {
    if (!has(key) || isNull(key)) {
        return null
    }
    val value = optLong(key, Long.MIN_VALUE)
    return value.takeIf { it != Long.MIN_VALUE }
}

private fun JSONObject.readOptionalString(key: String): String? {
    if (!has(key) || isNull(key)) {
        return null
    }
    return optString(key, "").ifBlank { null }
}
